fauna/query/iterator.py

```diff
@@ -0,0 +1,67 @@
+from typing import Iterator, Optional
+
+from fauna.client.client import Client, QueryOptions
+from fauna.query import Query, fql, Page
+
+
+class QueryIterator:
+    """A class to provider an iterator on top of Fauna queries."""
+
+    def __init__(self,
+                 client: Client,
+                 fql: Query,
+                 opts: Optional[QueryOptions] = None):
+        """Initializes the QueryIterator
+
+        :param fql: A string, but will eventually be a query expression.
+        :param opts: (Optional) Query Options
+
+        :raises TypeError: Invalid param types
+        """
+        if not isinstance(client, Client):
+            err_msg = f"'client' must be a Client but was a {type(client)}. You can build a " \
+                        f"Client by calling fauna.client.Client()"
+            raise TypeError(err_msg)
+
+        if not isinstance(fql, Query):
+            err_msg = f"'fql' must be a Query but was a {type(fql)}. You can build a " \
+                       f"Query by calling fauna.fql()"
+            raise TypeError(err_msg)
+
+        self.client = client
+        self.fql = fql
+        self.opts = opts
+        self.cursor = None
+
+    def __iter__(self) -> Iterator:
+        return self.iter()
+
+    def iter(self) -> Iterator:
+        initialResponse = self.client.query(self.fql, self.opts)
```

Comment:
In order to truly make this idempotent, maybe we should only ever query once, then set the page on self and reuse it in additional calls.

That also helps make this reusable as an embedded page iterator when we get there. e.g. we can implement Page.to_iter(client, opts) and init this object with the page already set.

What do you think?

Reply:
If a set comprises many pages, then we would need to store multiple pages, or are you suggestion only storing the first page? Then subsequent iterations will always use the same page to start, but who knows what subsequent pages would be.

Page.to_iter() would always suffer that effect of who-knows-what-subsequent-pages-will-hold, though the first page will always be itself of course.

hmmm...

I am not convinced that the iterator should be concerned with idempotence. It should be known to make external requests to fulfill the iterator.
- If what you want is to return the same documents each time, then you can structure your query to give you idempotence (e.g. provide the same index range).
- If what you want is to cache the results, then you can save copies of the data as you go to avoid the external requests. You can do that by stashing them somewhere of your choosing (with safeguards for memory consumption).

Reply:
For reusing, I am considering removing `fql` and `opts` from `__init__` and replacing with a stored iterator. Then we can have factory functions like `from_query` and `from_page`, each of which would defining the iterator to init the `QueryIterator` or `SetIterator` (would probably rename the iterator to be more generic if Page stuff was added). Then `Page.to_iter` would basically call `SetIterator.fromPage(client=client, page=self, opts=opts)`